src/engine/qcommon/common.cpp

```diff
@@ -861,26 +861,49 @@ void Com_Frame()
 		minMsec = 1;
 	}
 
+	int beforeProcessing = Sys::Milliseconds();
+
 	Com_EventLoop();
-	com_frameTime = Sys::Milliseconds();
 
-	if ( lastTime > com_frameTime )
-	{
-		lastTime = com_frameTime; // possible on first frame
-	}
+	// It must be called at least once.
+	IN_Frame();
 
-	msec = com_frameTime - lastTime;
+	int afterProcessing = Sys::Milliseconds();
+
+	com_frameTime = afterProcessing;
 
-	IN_Frame(); // must be called at least once
+	// lastTime can be greater than com_frameTime on first frame.
+	lastTime = std::min( lastTime, com_frameTime );
+
+	msec = com_frameTime - lastTime;
 
 	while ( msec < minMsec )
 	{
-		//give cycles back to the OS
-		Sys::SleepFor(std::chrono::milliseconds(std::min(minMsec - msec, 50)));
-		IN_Frame();
+		int processingTime = afterProcessing - beforeProcessing;
+
+		/* Fill as much as input frames as possible within this timeframe to not
+		let the game process events in a way it would last longer than that. */
+		int jamTime = std::max( afterProcessing, 50 );
```

Comment:
What's jam time? I'm not sure the comment is clear to me.

Reply:
The idea is that if you miss 50ms to reach the frametime, you should not sleep 50ms, because 1ms is not precise enough and if you stop sleeping 0.1ms after the frametime, you lost 1ms. And losing 1ms per frame makes the 1000fps cap becoming a 500fps one. So basically if you miss 50ms to reach the frametime, you better sleep for 45ms then loop on input until the frametime is reached, that non-sleeping final looping is what I call the “jam time”.

It's like you sleep for a whole night but wake up 10 min before the alarm clock is expected to ring, keeping the alarm clock in hand and checking the time running second per second, and you don't miss any second, so when the alarm clock rings, no one can beat you at getting up faster than you, and you would wake up that fast even if you were deaf. What I called the jam part is that part you wait for the clock to ring, while already being awake.

I will rewrite this part any way, actually the 50ms as a jam time means 20fps, so we currently never sleep. It's like, instead of waking up 10min before the clock ring, you keep yourself awake for the whole night. We only need to wake up 10 minutes before the alarm clock rings.

Reply:
The `50ms` came from the original code and that confused me as if this would be a valid value, even if it was only an upper limit for a sleep (which is now a lower limit for a jam), which is totally stupid in both cases, unless you're in 1999 and you expect most players to have less than 20 fps.